Create OpenVPN Access Server Helm chart (#1)
OpenVPN/as-helm-chart: README.md

```console
helm install my-vpn ./as-helm-chart
```

These commands deploy openvpn-as on the Kubernetes cluster in the default configuration.

## Uninstalling the Chart

To uninstall/delete the `my-vpn` deployment:

```console
helm delete my-vpn
```

The command removes all the Kubernetes components associated with the chart and deletes the release.

## Configuration

The configurable parameters of the openvpn-as chart can be found in `values.yaml`.

Specify each parameter using the `--set key=value[,key=value]` argument to `helm install`.

Alternatively, a YAML file that specifies the values for the parameters can be provided while installing the chart. For example,

```console
helm install my-vpn -f values.yaml ./as-helm-chart
```

## Access server custom configuration

Custom configuration is possible via Kubernetes [postStart](https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/) feature.

***Important: If you plan to use `sacli` in the `postStart` script, ensure that the Access Server service is fully up and running beforehand.***  
You can find an example of how to wait for the service in the Inline Script below.

> If enabled, this chart runs default `scripts/configure.sh` from the chart.

Reviewer: `If enabled` is pretty unclear here. Maybe something like "if you set `postStart.enabled=true`," ?

> This is just an example which waits while AS initialization is finished.

Reviewer: ```suggestion
However, that default script is just an example which waits while AS initialization is finished.
```

The default `scripts/configure.sh` can be overridden.

### Override with Inline Script

> 

Reviewer: Let's add something like "To override the postStart script you can provide the script directly via the values.yaml."

#### Add to values.yaml:

> ```console

Reviewer: ```suggestion
```yaml
```

postStart:
  customScript: |
    #!/bin/bash
    # Waiting for AS service initialization
    until /usr/local/openvpn_as/scripts/sacli status 2>/dev/null |grep -q '"api": "on"'
    do
        sleep 2
    done
    /usr/local/openvpn_as/scripts/sacli --user "openvpn" --new_pass "secure123" SetLocalPassword
```

> #### Use External Secret

Reviewer: Any specific reason you called this "External" secret? "External Secret" makes me think of https://external-secrets.io/, but you seem to just refer to a normal Kubernetes secret. What is "external" about it?

Reviewer: Ah, I understand it now. You create a secret from the script, so you call it external secret if the user provides it and you do not create it yourself. Maybe some explanation might help:

"The chart will create a Kubernetes secret out of the provided script. Instead you can also provide your own secret directly with the postStart.externalSecretName key. Note that the the script must be available as configure.sh in the secret."

Author: yeah, I agree this is confusing 
renamed to `secretRefName` and updated README

```console
postStart:
  externalSecretName: my-custom-script
```

Create it:

```console
kubectl create secret generic my-custom-script \
  --from-file=configure.sh=./custom.sh
```

## TUN device in Kubernetes

A TUN device in Kubernetes is a virtual network interface that allows for packet routing between user space and the kernel, commonly used for applications like VPNs. To use a TUN device, pods typically need specific permissions, and starting from Kubernetes version 1.31.3, they must run in privileged mode to create and access the TUN device.

To create and use TUN devices within Kubernetes pods, specific permissions are necessary:

> Kubernetes Versions Prior to 1.31.3: The NET_ADMIN capability was sufficient for creating TUN devices.

Reviewer: ```suggestion
Kubernetes versions prior to 1.31.3: The NET_ADMIN capability was sufficient for creating TUN devices.
```

Weird capitalization

> Kubernetes Version 1.31.3 and Later: Pods must run in privileged mode to create TUN devices. This change was made due to security improvements that restrict access to certain system resources.

Reviewer: ```suggestion
Kubernetes version 1.31.3 and later: Pods must run in privileged mode to create TUN devices. This change was made due to security improvements that restrict access to certain system resources.
```

> Also instead of using privileged mode the Device Plugins can be used

Reviewer: Aborted sentence?

```suggestion
Instead of using privileged mode [Device Plugins](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) can be used.
```

Kubernetes supports device plugins that can expose host devices to pods. For TUN devices, you can deploy a device plugin that allows access to /dev/net/tun. This can be done using a DaemonSet that runs on each node.

> Example Device Plugin Configuration

Reviewer: Was this intended to be a headline?

Here’s a basic example of a device plugin configuration:

> ```console

Reviewer: ```suggestion
```yaml
```

apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: tun-device-plugin
  namespace: kube-system
spec:
  template:
    spec:
      containers:
      - name: tun-device-plugin

>         image: your-device-plugin-image

Reviewer: Not sure whether this example is particularly useful. The actual work seems to be in creating the required code for the plugin. So

* we could check whether there is an existing OSS plugin for tun devices
* create our own
* if neither of those two, I would probably remove this example. Anyone who can write such a plugin can probably come up with the configuration on their own...

You can find an example of how to wait for the service in the Inline Script below.

If you set `postStart.enabled=true`, this chart runs default `scripts/configure.sh` from the chart.
However, that default script is just an example which waits while AS initialization is finished.
The default `scripts/configure.sh` can be overridden.

#### Override with Inline Script
To override the `postStart` script you can provide the script directly via the `values.yaml`.

```yaml
postStart:
  customScript: |
    #!/bin/bash
    # Waiting for AS service initialization
    until /usr/local/openvpn_as/scripts/sacli status 2>/dev/null |grep -q '"api": "on"'
    do
        sleep 2
    done
    /usr/local/openvpn_as/scripts/sacli --user "openvpn" --new_pass "secure123" SetLocalPassword
```

#### Use a separate Kubernetes Secret
The chart will create a Kubernetes secret out of the provided script. Instead you can also provide your own secret directly with the `postStart.secretRefName` key. Note that the the script must be available as `configure.sh` in the secret.
```yaml

> postStart:

Reviewer: I think this example should set `postStart.enabled`

  secretRefName: my-custom-script
```

Create it:

```console
kubectl create secret generic my-custom-script \
  --from-file=configure.sh=./custom.sh
```

## TUN device in Kubernetes

A TUN device in Kubernetes is a virtual network interface that allows for packet routing between user space and the kernel, commonly used for applications like VPNs. To use a TUN device, pods typically need specific permissions, and starting from Kubernetes version 1.31.3, they must run in privileged mode to create and access the TUN device.

To create and use TUN devices within Kubernetes pods, specific permissions are necessary:

> Kubernetes versions prior to 1.31.3: The NET_ADMIN capability was sufficient for creating TUN devices.

Reviewer: Looking at the formatted output in Github, this is not formatted very nicely. Maybe convert to list items? Or individual paragraphs? (Markdown doesn't generally seem to have `<dl>`, and a table seems overkill)

Author: missed that, convert to list
now looks better, wdyt?
